### ISSUE TITLE
Database tab is empty in generated docs

### ISSUE THREAD

**Issue by jesperbagge** [ClickHouse/dbt-clickhouse#311]
### Describe the bug
When serving the generated docs, I only have models under the Project tab. The Database tab is completely empty. Is this due to that dbt expects database.schema.table while ClickHouse skips the schema part?

Is this a concious choice or is my dbt version misconfigured?

### Steps to reproduce
1. run `dbt docs generate`
2. run `dbt docs serve`
3. Browse to the tab Overview -> Database

### Expected behaviour
I was expecting to look at trees of databases and tables.

#### Environment
* dbt version: 1.7.12
* dbt-clickhouse version: 1.7.6
* clickhouse-connect version (if using http): 0.7.0
* Python version: 3.11.9
* Operating system: macOS-14.5-arm64-arm-64bit

#### ClickHouse server
* ClickHouse Server version: 24.5.1.1763

**Comment by LLIyMAXEP** [comment]
@jesperbagge maybe you have figured out the causes of this issue or found a workaround?

**Comment by jesperbagge** [comment]
> [@jesperbagge](https://github.com/jesperbagge) maybe you have figured out the causes of this issue or found a workaround?

No, unfortunately not. I haven’t spent much time on this issue after reporting it.

**Comment by morsapaes** [comment]
I was expecting this to be a simple fix related to how we fetch catalog metadata, but it actually led me down a rabbit hole. AFAIU, there is a fundamental limitation in dbt Explorer ([`dbt-docs`](https://github.com/dbt-labs/dbt-docs)) that prevents rendering the `Database` view when the `database` passed is an empty string.

Because ClickHouse uses a different namespace hierarchy (_i.e._ not `database.schema.object`, but `database.object` masked as `schema.object` in the adapter), we hardcode the database as an empty string (`''`). Although the `dbt-docs` code seems prepared to handle this by falling back to `schema`, dbt core will serialize the empty string as `null` instead —  this causes a mismatch between the project and model-level metadata in `manifest.json`, which seems to be the underlying cause for the issue with rendering. You can verify this behavior manually using a simple repro:

```bash
dbt init test_docs

cd test_docs

dbt run # generates /target/manifest.json

dbt docs generate # generates /target/catalog.json

dbt docs serve # empty Database view
```

Open `manifest.json` and `catalog.json` and manually replace `"database": null` _and_ `"database": ""` with `"database":"test_docs"`, then run:

```bash
dbt docs serve
```

<img width="1216" alt="Image" src="https://github.com/user-attachments/assets/1081a551-7239-4b83-be0b-b9d70f0e7622" />

It's not straightforward to understand if or how the adapter could be patched, since the manifest metadata cannot be manipulated via the adapter layer. I tried setting a project-level `database` default, but that only partially helped by addressing the top-level `null` values. Maybe @dbeatty10 or someone working on `dbt-docs` has some ideas here!

**Comment by jesperbagge** [comment]
> ...we hardcode the database as an empty string (''). Although the dbt-docs code seems prepared to handle this by falling back to schema, dbt core will serialize the empty string as null instead...

Nice findings, @morsapaes ! Perhaps a silly question but what would break if we instead hardcoded the database into something generic, like the name of the project from `dbt_project.yaml` or simply `ClickHouse`? Just to get around dbtCore serializing empty strings as null.

**Comment by koletzilla** [comment]
I have been investigating a bit this to try to to get more knowledge on the issue and I have reached the same point as you folks.
Leaving two artifacts for future reference:
- #640 changes some pieces to try to have the `database` field filled. Looks simple, but tbh I don't trust this change at all. I'm afraid it will change the behavior of some part. With it the database is filled, but then the doc generation have naming inconsistencies in the same way as described here https://github.com/ClickHouse/dbt-clickhouse/issues/311#issuecomment-2955622423
- Script to instead of just randomly replacing `"database": ""`, do it by adding the same name as the one in `"schema"`. If some of you don't mind dealing with the naming of relations being now `the_database.the_database.the_table`, it may work to unblock you. Remember to run it between `dbt docs generate` and `dbt docs serve`
```python
#!/usr/bin/env python3
"""Patch target/manifest.json and target/catalog.json so that any
object with `"database": ""` gets its database set to the sibling
`"schema"` value.

Run from a dbt project root after `dbt docs generate`:

    python /path/to/fix_dbt_docs_db.py
"""

import json
import sys
from pathlib import Path


def patch(node):
    """Recursively replace `"database": ""` with the sibling `"schema"`
    value. Returns the number of patches applied."""
    n = 0
    if isinstance(node, dict):
        schema = node.get("schema")
        if node.get("database") == "" and isinstance(schema, str) and schema:
            node["database"] = schema
            n += 1
        for v in node.values():
            n += patch(v)
    elif isinstance(node, list):
        for v in node:
            n += patch(v)
    return n


target = Path.cwd() / "target"
if not target.is_dir():
    sys.exit(f"error: no `target/` directory in {Path.cwd()}. Run `dbt docs generate` first.")

total = 0
for name in ("manifest.json", "catalog.json"):
    path = target / name
    if not path.exists():
        print(f"  skip: {name} (not found)")
        continue
    data = json.loads(path.read_text(encoding="utf-8"))
    n = patch(data)
    if n:
        path.write_text(json.dumps(data, indent=4, ensure_ascii=False) + "\n", encoding="utf-8")
        print(f"  patched {n} object(s) in {name}")
    else:
        print(f"  no changes in {name}")
    total += n

print(f"done. patched {total} object(s) total.")

```
